oioswift/common/middleware/container_sharding.py

```diff
@@ -0,0 +1,534 @@
+        pipeline = str(PipelineWrapper(ctx)).split(' ')
+
+        if 'swift3' in pipeline and not all((self.account_first,
+                                             self.strip_v1,
+                                             self.swift3_compat)):
+            LOG.warn('account_first, strip_v1 and swift3_compat options '
+                     'must be enabled when using %s along with swift3',
+                     MIDDLEWARE_NAME)
+
+        auth_index = -1
+        if 'tempauth' in pipeline:
+            LOG.debug('Use tempauth middleware.')
+            auth_index = pipeline.index('tempauth')
+        elif 'keystoneauth' in pipeline:
+            LOG.debug('Use keystone middleware.')
+            auth_index = pipeline.index('keystoneauth')
+        if pipeline.index(MIDDLEWARE_NAME) < auth_index:
+            raise ValueError(
+                'Invalid pipeline %r: %s must be placed after authentication'
+                % (pipeline, MIDDLEWARE_NAME))
+
+        if ('slo' in pipeline and
+                pipeline.index(MIDDLEWARE_NAME) < pipeline.index('slo')):
+            raise ValueError(
+                'Invalid pipeline %r: %s must be placed after SLO'
+                % (pipeline, MIDDLEWARE_NAME))
+
+    def key(self, account, container, mode, path=None):
+        """returns Redis key"""
+        ret = self.PREFIX + account + ":" + container + ":"
+        if mode:
+            ret += mode + ":"
+            if path:
+                ret += path
+        return ret
+
+    def _create_key(self, req, account, container, mode, path):
+        key = self.key(account, container, mode, path)
+        LOG.debug("%s: create key %s", self.SWIFT_SOURCE, key)
+        # TODO: should we increase number of objects ?
+        # but we should manage in this case all error case
+        # to avoid false counter
+        res = self.conn.set(key, "1")
+        if not res:
+            LOG.warn("%s: failed to create key %s", self.SWIFT_SOURCE, key)
+
+    def _remove_key(self, req, account, container, mode, path):
+        key = self.key(account, container, mode, path)
+        if mode == CNT:
+            # remove container key only if empty
+            empty = not any(self._list_objects(
+                            req.environ.copy(),
+                            account,
+                            [container] + path.split('/')[:-1],
+                            None,
+                            limit=1))
+
+            if not empty:
+                return
+
+        LOG.debug("%s: remove key %s (key %s)", self.SWIFT_SOURCE, path, key)
+        res = self.conn.delete(key)
+        if not res:
+            LOG.warn("%s: failed to remove key %s", self.SWIFT_SOURCE, key)
+
+    def _build_object_listing(self, start_response, env,
+                              account, container, prefix,
+                              limit=None,
+                              recursive=False, marker=None):
```

Comment:
This method is very long, could we split it?

Reply:
Will be done in dedicated PR